### PR TITLE
Fix LoRA freeze logic: mm_tunable_parts incorrectly unfreezes LLM backbone

### LINKED ISSUE #508: LoRA Logic Bug Report:mm_tunable_parts + LoRA = Full-Model Training (Unintended)
> **Preface**
> Thanks for maintaining this project. I’m sharing my findings in good faith while reproducing EgoBlind. If I’ve misunderstood your original intent, design choices, or any part of the code logic, please forgive me—and I’d really appreciate any correction. I’m happy to adjust the report/PR accordingly once I understand the intended behavior.
### LoRA’s wrong logic in LLaVA-NeXT

When I was running the [EgoBlind](https://github.com/doc-doc/EgoBlind) dataset on the LLaVA-NeXT model (which needs to train the MLP layer), I found that after using LoRA fine-tuning, the overall VRAM usage (about 160G) and the inference speed almost did not change at all! Previously, someone in the issues of your project’s GitHub repository mentioned a similar problem [issues#253](https://github.com/LLaVA-VL/LLaVA-NeXT/issues/253). This is obviously inconsistent with the general phenomenon after using LoRA fine-tuning. By reading through all relevant code in the LLaVA-NeXT repository, I found a major error in the LoRA logic in LLaVA-NeXT! Its specific manifestation is: as long as the original train.py code uses the mm_tunable_parts parameter, it is impossible to perform LoRA fine-tuning in the true sense (i.e., freezing the backbone model and only fine-tuning the LoRA parameters)!!! And if you do not use the mm_tunable_parts parameter, it is forever impossible to achieve true LoRA fine-tuning + training the MLP layer (or vision_resampler, vision_tower)! Specifically as follows:

#### Using the mm_tunable_parts parameter

First, it should be clarified that if you want to use LoRA fine-tuning, inevitably the lora_enable parameter will be set to True, and LoRA will be injected into the model first.

If you use the `mm_tunable_parts` parameter, naturally you will enter the else branch of this code block (line 1615) in trian.py:

```python
if model_args.mm_tunable_parts is None:  # traditional way of deciding which part to train
    model.config.tune_mm_mlp_adapter = training_args.tune_mm_mlp_adapter = model_args.tune_mm_mlp_adapter
    model.config.tune_mm_vision_resampler = training_args.tune_mm_vision_resampler = model_args.tune_mm_vision_resampler
    if model_args.tune_mm_mlp_adapter or model_args.tune_mm_vision_resampler:
        model.requires_grad_(False)
    if model_args.tune_mm_mlp_adapter:
        for p in model.get_model().mm_projector.parameters():
            p.requires_grad = True
    if model_args.tune_mm_vision_resampler:
        for p in model.get_model().vision_resampler.parameters():
            p.requires_grad = True

    model.config.freeze_mm_mlp_adapter = training_args.freeze_mm_mlp_adapter
    if training_args.freeze_mm_mlp_adapter:
        for p in model.get_model().mm_projector.parameters():
            p.requires_grad = False

    model.config.freeze_mm_vision_resampler = training_args.freeze_mm_vision_resampler
    if training_args.freeze_mm_vision_resampler:
        for p in model.get_model().vision_resampler.parameters():
            p.requires_grad = False

    model.config.unfreeze_mm_vision_tower = model_args.unfreeze_mm_vision_tower
    if model_args.unfreeze_mm_vision_tower:
        vision_tower.requires_grad_(True)
    else:
        vision_tower.requires_grad_(False)

else:
    rank0_print(f"Using mm_tunable_parts: {model_args.mm_tunable_parts}")
    model.config.mm_tunable_parts = training_args.mm_tunable_parts = model_args.mm_tunable_parts
    # Set the entire model to not require gradients by default
    model.requires_grad_(False)
    vision_tower.requires_grad_(False)
    model.get_model().mm_projector.requires_grad_(False)
    model.get_model().vision_resampler.requires_grad_(False)
    # Parse the mm_tunable_parts to decide which parts to unfreeze
    tunable_parts = model_args.mm_tunable_parts.split(",")
    if "mm_mlp_adapter" in tunable_parts:
        for p in model.get_model().mm_projector.parameters():
            p.requires_grad = True
    if "mm_vision_resampler" in tunable_parts:
        for p in model.get_model().vision_resampler.parameters():
            p.requires_grad = True
    if "mm_vision_tower" in tunable_parts:
        for name, param in model.named_parameters():
            if "vision_tower" in name:
                param.requires_grad_(True)
    if "mm_language_model" in tunable_parts:
        for name, param in model.named_parameters():
            if "vision_tower" not in name and "mm_projector" not in name and "vision_resampler" not in name:
                param.requires_grad_(True)
```

After entering the else branch, it will first freeze all parts (including the injected LoRA), and then precisely unfreeze the specified parts according to mm_tunable_parts. When "mm_language_model" does not exist in mm_tunable_parts, LoRA will always be frozen; this is correct, after all, the backbone model is not trained. But then the problem comes: if "mm_language_model" exists in mm_tunable_parts, then it will enter the `if "mm_language_model" in tunable_parts:` sub-branch. At this time, it will unfreeze all parameters whose names do not contain vision_tower/mm_projector/vision_resampler, **including LoRA**! Then the bizarre thing comes: the LoRA parameters will be fine-tuned together with mm_language_model (i.e., the backbone model)! And we all know that the true meaning of LoRA is to freeze the backbone model and only fine-tune the LoRA parameters; doing this is obviously not the correct LoRA way!

#### Not using the mm_tunable_parts parameter

If you do not use the mm_tunable_parts parameter, it is indeed possible to achieve true LoRA fine-tuning. For example, set the mm_tunable_parts parameter to none, set the tune_mm_mlp_adapter parameter to False (why tune_mm_mlp_adapter is mentioned will be said later), and only set the lora_enable parameter to True. After my review, it will follow PEFT’s default behavior; generally it makes only the LoRA parameters trainable and freezes the backbone weights. At this time, it can indeed achieve true LoRA fine-tuning!

However, this fine-tuning method, besides the LoRA parameters, will freeze all other parts of the model (mm_mlp_adapter, vision_resampler, vision_tower), that is, it can only fine-tune LoRA. This is extremely, extremely inflexible! This is also absolutely not the result we want!

Furthermore, LLaVA-NeXT/scripts/train/finetune_ov.sh is an example fine-tuning script, in which the mm_tunable_parts parameter is clearly included. Ordinary researcher users will generally not, like me, read the code so carefully. When they want LoRA fine-tuning, generally they will directly set `--mm_tunable_parts="mm_mlp_adapter,mm_language_model"`, `--lora_enable True` (obviously the EgoBlind authors and issues#253 did it this way). They mistakenly think they used LoRA fine-tuning, but in fact not. In the end, compared to the previous full fine-tuning, they even fine-tuned the LoRA parameters additionally! VRAM usage and inference speed almost did not change; this is undoubtedly a misunderstanding and abuse of LoRA fine-tuning!

So I think the code in the `if "mm_language_model" in tunable_parts:` sub-branch in the above code block can be succinctly changed to the following code (I have already uploaded a Pull requests):

```python
if "mm_language_model" in tunable_parts:
    if training_args.lora_enable:
        rank0_print("LoRA enabled: Unfreezing ONLY LoRA parameters in LLM.")
        for name, param in model.named_parameters():
            if "lora_" in name:
                param.requires_grad_(True)
    else:
        rank0_print("LoRA disabled: Unfreezing FULL LLM backbone.")
        for name, param in model.named_parameters():
            if "vision_tower" not in name and "mm_projector" not in name and "vision_resampler" not in name and "mm_depth_projector" not in name:
                param.requires_grad_(True)
```

#### Test results

After my tests on the EgoBlind dataset, the training VRAM usage is about 56G after modification (originally close to 160G), the inference time is reduced from the original about 7 hours to 4.5 hours, fully in line with the expected effect of LoRA fine-tuning! The original fine-tuned model weight file was about 12GB, and the modified model weight file was about 2GB, also in line with the small-model expectation of LoRA fine-tuning! None of these phenomena fails to show that my analysis of the LoRA logic error in LLaVA-NeXT is correct! On the public EgoBlind dataset, the accuracy also increased from 57% to around 59%, and the effect improved to some extent!The modified code can perfectly adapt to the scripts in the repository; in fact, after I modified the code, it ran successfully in one go.
Sorry, I didn’t save the runtime time or VRAM usage data. Here I only attach comparison images of the model file size.

Image

The archive above is the model file obtained before the code change, with a size of 11.31 GB; the archive below is the model file obtained after the code change, with a size of 88.52 MB.
> When checking the code of this repository, I found that the tune_mm_mlp_adapter parameter is used to separately choose whether to fine-tune the MLP layer: the tune_mm_mlp_adapter parameter defaults to False, but if it is set to True, then except for the MLP layer, all other parts of the model will be frozen, including the LoRA parameters (even if lora_enable is set to True), that is, only the MLP layer is trained. [issues#438](https://github.com/LLaVA-VL/LLaVA-NeXT/issues/438)

### PR DESCRIPTION
**Pull request by Nyquist24**
## Summary

This PR fixes the LoRA parameter freeze/unfreeze logic when used together with `mm_tunable_parts` in both `train.py` and `train_dpo.py`.

**Issue:**  
When `--lora_enable True` and `--mm_tunable_parts` includes `mm_language_model`, the code was unfreezing **all** LLM backbone parameters instead of just LoRA weights, effectively turning LoRA fine-tuning into full fine-tuning. On the other hand, if `mm_language_model` was not included, LoRA parameters ended up frozen by `model.requires_grad_(False)` and never restored.

**Impact:**  
Users who configured `--mm_tunable_parts="mm_mlp_adapter,mm_language_model" --lora_enable True`—a common setup—were unknowingly performing full fine-tuning, which increased VRAM usage by ~3x and removed the intended LoRA efficiency benefits.

## Changes

- **`mm_language_model` branch:** Correctly distinguish between LoRA mode (unfreeze only LoRA parameters) and full fine-tune mode (unfreeze the entire LLM backbone).  
- **LoRA parameter restoration:** After any global `model.requires_grad_(False)`, LoRA parameters (`lora_*`) are explicitly restored to trainable. This applies to both the `mm_tunable_parts` branch and the legacy `tune_mm_mlp_adapter` branch.  
- **Warnings:** Log a warning when `lora_enable=True` but `mm_language_model` is not in `mm_tunable_parts`.  
- **Trainable parameter summary:** Log per-group parameter counts (LoRA, `mm_projector`, `vision_resampler`, `vision_tower`, `llm_base`) at startup to make verification easier.  
- **Files updated:** The fix is applied consistently to both `train.py` (SFT) and `train_dpo.py` (DPO).

## Before vs After

| Configuration | Before (bug) | After (fixed) |
|---|---|---|
| `mm_tunable_parts="mm_mlp_adapter,mm_language_model"` + `lora_enable=True` | LoRA + MLP + **entire LLM backbone** | LoRA + MLP only |
| `mm_tunable_parts="mm_mlp_adapter"` + `lora_enable=True` | MLP only (LoRA frozen) | LoRA + MLP |
| `mm_tunable_parts=None` + `tune_mm_mlp_adapter=True` + `lora_enable=True` | MLP only (LoRA frozen) | LoRA + MLP |

## Test Plan

- Run with `--mm_tunable_parts="mm_mlp_adapter,mm_language_model" --lora_enable True` and confirm log shows `llm_base=0` and `lora>0`.  
- Run with `--mm_tunable_parts="mm_mlp_adapter,mm_language_model" --lora_enable False` and confirm `llm_base>0` (full fine-tune unchanged).  
- Run with `--mm_tunable_parts="mm_mlp_adapter" --lora_enable True` and confirm both `lora>0` and `mm_projector>0`.  
- Verify VRAM usage is significantly lower in LoRA mode compared to before.

Fixes #508  
Related: #253, #438

### CODE REVIEW

**Comment by lepelo0811** [comment]
Hi, thanks for working on this fix.
I wanted to mention that this issue was originally identified from my side, and I spent quite a lot of time reproducing it, narrowing down the incorrect freeze/unfreeze behavior, and evaluating which parameter groups were actually being trained under different configurations. Recently I have been quite busy, so I did not have the time to finish a complete patch myself.
Your fix looks good, and I appreciate that you pushed it forward. At the same time, I can clearly see that parts of the problem analysis and verification logic are closely related to what I had already posted in the issue/comments.
Would you mind adding me as a co-author in the relevant commit message(s)? I would really appreciate proper attribution. Otherwise, if this PR gets merged without listing me as a co-author, all the effort I put into discovering, analyzing, and validating this issue would unfortunately not be properly reflected.
You can add the following trailer to the commit message:
Co-authored-by: lepelo0811 <2024218328@mail.hfut.edu.cn>
If you are not familiar with how to do this, you can find the official GitHub instructions here:
https://docs.github.com/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors
If there are any questions, you can also reach me through this email.
Thanks again.

**Comment by Nyquist24** [comment]
> Hi @lepelo0811 , I've added you as a co-author in the commit. Thanks for your contribution!

**Comment by lepelo0811** [comment]
> > Hi @lepelo0811 , I've added you as a co-author in the commit. Thanks for your contribution!

Thank u very much!

**Review by kcz358 — APPROVED** [review]

**Comment by Luodian** [comment]
I re-checked this against current `main` and the freeze/unfreeze fix here is the right direction.

There is one remaining follow-up outside `train.py` / `train_dpo.py`: in `llava/train/llava_trainer.py`, the adapter-only checkpoint branch still runs for single-part adapter tuning even when `lora_enable=True`. That means configs like `--mm_tunable_parts="mm_mlp_adapter" --lora_enable True` can train correctly in memory after this fix, but still save adapter-only checkpoints that omit LoRA weights for resume.

I opened #511 to carry this PR forward and close that checkpoint/save gap. It includes this PR commit unchanged plus a small follow-up in `llava_trainer.py`.

If you prefer to keep this PR as the merge target, cherry-picking commit `bb42acc` on top of #510 should address the remaining issue.

**Comment by lepelo0811** [comment]
> I re-checked this against current `main` and the freeze/unfreeze fix here is the right direction.
> 
> There is one remaining follow-up outside `train.py` / `train_dpo.py`: in `llava/train/llava_trainer.py`, the adapter-only checkpoint branch still runs for single-part adapter tuning even when `lora_enable=True`. That means configs like `--mm_tunable_parts="mm_mlp_adapter" --lora_enable True` can train correctly in memory after this fix, but still save adapter-only checkpoints that omit LoRA weights for resume.
> 
> I opened #511 to carry this PR forward and close that checkpoint/save gap. It includes this PR commit unchanged plus a small follow-up in `llava_trainer.py`.
> 
> If you prefer to keep this PR as the merge target, cherry-picking commit `bb42acc` on top of #510 should address the remaining issue.

It seems I don't have edit access to this PR. I am currently trying to contact @Nyquist24. Thank you for your recognition and understanding. Please bear with us for a little while.

**Comment by Luodian** [comment]
Maintainer note: the complete fix was merged via #511.

Thank you to @Nyquist24 for the broader freeze/unfreeze fix and to @lepelo0811 for identifying, analyzing, validating, and co-authoring the original fix path. The merged solution preserved the original fix commit from this PR and added the small checkpoint/save follow-up needed for complete LoRA + adapter support.
